### PR TITLE
[FLINK-25046][runtime] Convert unspecified edge to rescale when using adapive batch scheduler

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change

Currently, unspecified edges(partitioner == null) will be setted to FORWARD when the parallelism of both upstream and downstream is -1. This may cause many job vertices whose parallelism is not calculated based on data volume, but is aligned with their upstream vertices' parallelism because of forward edges, which is contrary to the original intention of our adaptive batch scheduler.

For this problem, we will convert the unspecfied edges between OperatorChains to RESCALE edges(instead of FORWARD edges) after OperatorChains' creation.

## Verifying this change
UT `ForwardForUnspecifiedPartitionerTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/partitioner/ForwardPartitionerTestBase.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.flink.api.common.RuntimeExecutionMode;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.streaming.api.transformations.PartitionTransformation;
+import org.apache.flink.util.TestLogger;
+
+/** Base class for test {@link ForwardPartitioner}'s subclass. */
+public class ForwardPartitionerTestBase extends TestLogger {
```
Comment (zhuzhurk) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/partitioner/ForwardPartitionerTestBase.java:29]: This class only contains methods that can be static. It's better to rework it to be a `PartitionerTestUtils`.

**Review by zhuzhurk — COMMENTED** [review]
Nice work! @wanglijie95 
The changes looks good to me except for one minor comment.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/partitioner/ForwardPartitionerTestBase.java (hunk 1 above)
Comment (wanglijie95) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/partitioner/ForwardPartitionerTestBase.java]: Fixed

**Comment by wanglijie95** [comment]
@zhuzhurk Thanks for review. I 've addressed your comment. Please help to review it again.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by zhuzhurk** [comment]
Merging.

**Pull request closed without merging**